pkg/bgpv1/manager/reconciler/service.go

```diff
@@ -298,31 +293,61 @@ func (r *LBServiceReconciler) svcDesiredRoutes(newc *v2alpha1api.CiliumBGPVirtua
 	if !svcSelector.Matches(serviceLabelSet(svc)) {
 		return nil, nil
 	}
-
-	// Ignore service managed by an unsupported LB class.
-	if svc.Spec.LoadBalancerClass != nil && *svc.Spec.LoadBalancerClass != v2alpha1api.BGPLoadBalancerClass {
-		// The service is managed by a different LB class.
-		return nil, nil
-	}
-
 	// Ignore externalTrafficPolicy == Local && no local endpoints.
 	if svc.Spec.ExternalTrafficPolicy == slim_corev1.ServiceExternalTrafficPolicyLocal &&
 		!hasLocalEndpoints(svc, ls) {
 		return nil, nil
 	}
-
 	var desiredRoutes []netip.Prefix
-	for _, ingress := range svc.Status.LoadBalancer.Ingress {
-		if ingress.IP == "" {
-			continue
-		}
-
-		addr, err := netip.ParseAddr(ingress.IP)
-		if err != nil {
-			continue
+	// Loop over the service advertisements and determine the desired routes.
+	for _, svcAdv := range newc.ServiceAdvertisements {
+		switch svcAdv {
+		case v2alpha1api.BGPLoadBalancerIPAddr:
+			if svc.Spec.Type != slim_corev1.ServiceTypeLoadBalancer {
+				continue
+			}
+			// Ignore service managed by an unsupported LB class.
+			if svc.Spec.LoadBalancerClass != nil && *svc.Spec.LoadBalancerClass != v2alpha1api.BGPLoadBalancerClass {
+				// The service is managed by a different LB class.
+				return nil, nil
+			}
+			for _, ingress := range svc.Status.LoadBalancer.Ingress {
+				if ingress.IP == "" {
+					continue
+				}
+				addr, err := netip.ParseAddr(ingress.IP)
+				if err != nil {
+					continue
+				}
+				desiredRoutes = append(desiredRoutes, netip.PrefixFrom(addr, addr.BitLen()))
+			}
+		case v2alpha1api.BGPClusterIPAddr:
+			if svc.Spec.Type != slim_corev1.ServiceTypeClusterIP {
+				continue
+			}
+			if svc.Spec.ClusterIP == "" || len(svc.Spec.ClusterIPs) == 0 {
+				return nil, nil
+			}
+			if svc.Spec.ClusterIP != "" {
+				addr, err := netip.ParseAddr(svc.Spec.ClusterIP)
+				if err != nil {
+					continue
+				}
+				desiredRoutes = append(desiredRoutes, netip.PrefixFrom(addr, addr.BitLen()))
+			}
+			for _, clusterIP := range svc.Spec.ClusterIPs {
+				if clusterIP == "" {
+					continue
+				}
+				addr, err := netip.ParseAddr(clusterIP)
+				if err != nil {
+					continue
+				}
+				desiredRoutes = append(desiredRoutes, netip.PrefixFrom(addr, addr.BitLen()))
+			}
+		case v2alpha1api.BGPExternalIPAddr:
+			// TODO implement
```

Comment:
If you don't intend to implement  `ExternalIP` in this PR, I think it would be good idea to at least log a warning for the `ExternalIP` advertisement not being implemented, since it will be possible to configure it as of the CRD change in this PR.

Reply:
I don't know how to prevent users from setting `BGPExternalIPAddr`. Should I just delete it? @YutaroHayakawa 


https://github.com/cilium/cilium/blob/b463146191641d760a368cad3a5bf6155f921ee2/pkg/k8s/apis/cilium.io/v2alpha1/bgp_advert_types.go#L57
 or https://github.com/cilium/cilium/blob/b463146191641d760a368cad3a5bf6155f921ee2/pkg/k8s/apis/cilium.io/v2alpha1/bgp_advert_types.go#L37

```
// +kubebuilder:validation:Enum=LoadBalancerIP;ClusterIP;
```

Reply:
You can delete it for now and introduce it again in the PR when we introduce ExternalIP support. It is not a breaking change.

Reply:
Done.